### DISCUSSION TITLE
Additional details related to Trojan:PowerShell/Mountsi.A!ml

### DISCUSSION THREAD

**Discussion by potatoqualitee** [PowerShell/PowerShell#15415] [General]
## Summary of the new feature/enhancement

As a trusted community module maintainer, I would like a wealth of information shouted from the rooftops about how to prevent PowerShell scripts from being reported as a trojan by Windows Defender. Recently, a file that has not changed in 2 years was reported as a Trojan to two of our users.

![image](https://user-images.githubusercontent.com/8278033/118113836-14a35d80-b3e7-11eb-85ff-5cfecc670245.png)

While our file is not a trojan, false positives from Windows Defender impact the trustworthiness of our module. We've worked hard over the years to engender trust in both dbatools and PowerShell. Currently, I hold up our CI/CD process to review each commit to ensure it is not malicious before I publish it to the PowerShell Gallery.

Is there a solution for whitelisting trusted community modules and adding a feature to PSScriptAnalyzer that highlights problematic techniques?

So far, one user has confirmed that the signed version from the PowerShell Gallery is not triggering AV. The problem with this is that:

1. Not all PowerShell Developers can afford Code Signing Certificates
2. Not everyone uses the PowerShell Gallery and we'd like to continue offering installs via GitHub. Our code cannot be signed and saved to the repository.

It seems that this may be a result of some enhancements to [AMSI and machine learning](https://www.microsoft.com/security/blog/2020/08/27/stopping-active-directory-attacks-and-other-post-exploitation-behavior-with-amsi-and-machine-learning/)

> Antimalware Scan Interface (AMSI) helps security software to detect such malicious scripts by exposing script content and behavior. AMSI integrates with scripting engines on Windows 10 as well as Office 365 VBA to provide insights into the execution of PowerShell, WMI, VBScript, JavaScript, and Office VBA macros. 

(I hope IronPython is impacted as well 😅)

Here's information about the Defender definition that alerted the user

![image](https://user-images.githubusercontent.com/8278033/118118685-8bdbf000-b3ed-11eb-94f2-d73681a7f79d.png)

I assume that `[Reflection.Assembly]::LoadFrom($assemblyPath)`, which we use to make imports as fast as possible, is the problematic technique, but I can't find any confirmation. Other blogs talk about using that for loading from memory but we are loading from disk. Once we get this issue solved, however, I'd like to know what to modify in other files. Perhaps the Gallery can detect these techniques and email module owners.

Other PowerShell users are experiencing this a well as noted in the following links:

https://twitter.com/JustinWGrote/status/1392518022900850688
https://twitter.com/psCookieMonster/status/1392796401545981953
https://twitter.com/DrAzureAD/status/1361298815815417856
https://twitter.com/PrzemyslawKlys/status/1392933835978027019
https://github.com/PowerShell/vscode-powershell/issues/3017
https://www.reddit.com/r/PowerShell/comments/jj4wzw/removing_trojanpowershellmountsiaml_trigger_from/
https://techcommunity.microsoft.com/t5/microsoft-defender-for-endpoint/defender-detected-powershell-ise-exe-as-trojan-powershell/m-p/2180534

In addition, [this article](https://web.archive.org/web/20210421190712/https://getimageview.net/2020/06/02/microsoft-defender-smartscreen-is-hurting-independent-developers/) suggests that the file being unsigned is the problem, and a commenter said that "Cleared it up by submitting it to Microsoft’s service at https://www.microsoft.com/en-us/wdsi/filesubmission." Do we make it part of our GitHub pipeline to submit changed files to that site?

Edit: I know that the PowerShell team is not the Defender team,  but considering unsigned PowerShell files in a git repository (it's impossible to sign and commit, I've tried, it changes the SHA) trigger AV warnings, this may be a very big problem for PowerShell development and scripting in general. Taking a look at [Warren](https://twitter.com/psCookieMonster/status/1392796401545981953)'s code, which has been flagged by Defender, nothing stands out.

```powershell
# Pester wasn't mocking git... 
# Borrowed idea from https://github.com/pester/Pester/issues/415
function Invoke-ExternalCommand {
    [cmdletbinding()]
    param($Command, [string[]]$Arguments, [switch]$Passthru)

    Write-Verbose "Running $Command with arguments $($Arguments -join "; ")"
    $result = $null
    $result = & $command @arguments  
    Write-Verbose "$($result | Out-String)"
    if($Passthru)
    {
        $Result
    }
}
````

**Comment by tig** [comment]
I don't think this is specific to your script.

Defender has quarantined my `profile.ps1`! 

![image](https://user-images.githubusercontent.com/585482/118169474-d5bcda00-b3dd-11eb-94a0-0cda73de283d.png)


This is insane.

**Comment by potatoqualitee** [comment]
Crushing, @tig! Hopefully this gets fixed soon.

It was suggested that I ping @SteveL-MSFT and @PaulHigin to ask if they can get the MS Defender team to scan the GitHub Repos of popular community modules help teach the ML.

**Comment by SteveL-MSFT** [comment]
The Defender team continuously refines and update their heuristics to detect potential malicious scripts.  I don't think there's a way to validate one moment in time and ensure it'll always pass detection (as the rules keep changing/improving) with the exception of signing it with the Windows certificate (yeah, that's not going to happen!).

I'm pretty sure the Defender team already uses GitHub Open Source code as a resource, but I don't think they can just give any projects a free pass as you can't guarantee that a well known and popular piece of code doesn't get attacked and malicious code gets added.  See [this article](https://thenewstack.io/github-open-source-projects-entangled-by-the-octopus-malware-scanner/) as one example.

**Comment by potatoqualitee** [comment]
Thank you for the detailed response. Is there any way you can engage the Defender team to get a confirmation either way?

**Comment by tig** [comment]
> The Defender team continuously refines and update their heuristics to detect potential malicious scripts. I don't think there's a way to validate one moment in time and ensure it'll always pass detection (as the rules keep changing/improving) with the exception of signing it with the Windows certificate (yeah, that's not going to happen!).
> 
> I'm pretty sure the Defender team already uses GitHub Open Source code as a resource, but I don't think they can just give any projects a free pass as you can't guarantee that a well known and popular piece of code doesn't get attacked and malicious code gets added. See [this article](https://thenewstack.io/github-open-source-projects-entangled-by-the-octopus-malware-scanner/) as one example.

But @SteveL-MSFT, the fact that my PROFILE ps1 gets literally deleted by Defender is way different than Defender flagging ps1s that are actually found in the wild.

I've had to disable Defender's scanning for this vuln to keep it from deleting my PowerShell profile! 

I can't believe I'm the only person this is happening to!

**Comment by regner** [comment]
> I can't believe I'm the only person this is happening to!

You're not. Our TeamCity builds started failing across the board two days ago after a Windows 10 security update. To get builds running again we had to disable real-time protection. Which was deemed acceptable in the short-term while IT works on bringing on a new AV system that is not Windows Defender.

**Comment by gsaluja9** [comment]
I saw this too. any recommended actions ?
![psvirus](https://user-images.githubusercontent.com/52312085/118289904-70c7b980-b49b-11eb-9280-8efd6d2f09b1.jpg)

**Comment by jt-eic** [comment]
For the record, I'm getting this issue as well.

**Comment by mackposh** [comment]
@SteveL-MSFT I was just woken up at 3am this morning by alerting because a script we run on a nightly basis on ~10,000 endpoints Defender decided suddenly was a Trojan. This script hasn’t changed in months. These endpoints are in a PCI compliant segregated environment with no domain. This is a real problem. What’s the point of using a particular AV product if the ratio of false positives to legitimate events is high? At some point, it becomes the boy crying wolf.

**Comment by potatoqualitee** [comment]
Btw, Steve said this on Twitter 

> Some folks are getting false positives from Windows Defender recently with their legitimate PowerShell scripts.  We've contacted someone on the Defender team, so hopefully this gets resolved quickly.

**Comment by mackposh** [comment]
First, apologies to @SteveL-MSFT if my last comment was a little salty. I realize this isn't Powershell's problem directly. For some additional info, the script in question hasn't been touched in 2 years, and after reviewing it looking for anything we were doing that might have triggered AMSI(like direct .Net calls or inline C# type of things), it's as vanilla Powershell as you can probably get. That said, the irony of it all is that this is the script we use to make sure certain file extensions and directories are in Defender's exclusion lists. We are temporarily disabling certain Defender behaviors like real-time network scanning.

I can see why heuristics may have interpreted that behavior as potentially malicious, but if you can't manage Defender policies with the Defender Powershell module then what's the alternative? Again, for our environment, these endpoints are not in a domain and have no network access except for what we specifically allow, so solutions like Group Policy are not an option.

It would be preferable if Defender had options to customize actions to take for different events. In this case, warn us, but don't quarantine immediately. Understandably, from a security perspective, that increases the potential risk for us, especially if a specific event is actually legitimately malicious. I'd argue though that that is something that we, as an organization, should have processes and solutions in place to handle such an event. 

In our case, due to the segregated nature of our network environment, potential impact is rather low. We can simply replace the affected endpoint(s) without any worry of the compromise being able to spread to other systems, so for us, such a risk may be acceptable. It wouldn't for everyone though, so this is a tricky problem to solve.

Still, I think the problem is that Defender's heuristics are a wee bit _too sensitive_ and need dialed back a bit.

**Comment by TravisEz13** [comment]
WG-Security

PowerShell must follow the AMSI result.  If we add ability to alter how we respond, it becomes an attack vector to AMSI plug-ins.  We would need an AMSI feature that tells us that we could, say warn, instead of block. This would probably be by return code from AMSI.

Otherwise, you need to work through [Microsoft Security Intelligence](https://www.microsoft.com/en-us/wdsi/filesubmission) to get the file cleared as not malware.

I think we can continue this as a discussion.

**Reply by mackposh** [reply]
The point is though that vanilla Powershell _shouldn't_ be getting flagged in the first place by AMSI. I may be interpreting you wrong, and I apologize if I am, but your comment along with a few others I've seen this week about this, almost suggest that submitting a script to Microsoft to be verified becomes a part of any release cycle. Forgive me if I find that step a bit onerous and should be unnecessary. 

Now, if we were in-lining a bunch of custom C# or making more direct .Net calls than Powershell, I could see the need for that, but I'd wager the majority of Powershell scripts in regular use won't fit into either category. 

I do understand what you are saying otherwise; that is, currently the only real option if a person gets a legitimate script flagged by AMSI is to submit it for verification.

**Comment by tig** [comment]
@TravisEz13, if I understand you correctly, you are asserting that every Powershell user on the planet, needs to individually get their `$profile` files cleaned as not malware.

**Reply by potatoqualitee** [reply]
Something was dialed up. Has it been dialed back down? When should we expect this to be resolved? 

Just read on Twitter:

> I’ve been informed that the classifier generating the false positives has currently been disabled while the security research team is figuring how to fix it
